SIPS/sip-64.md

```diff
@@ -0,0 +1,141 @@
+
+## Specification
+
+<!--The specification should describe the syntax and semantics of any new feature, there are five sections
+1. Overview
+2. Rationale
+3. Technical Specification
+4. Test Cases
+-->
+
+### Overview
+
+<!--This is a high level overview of *how* the SIP will solve the problem. The overview should clearly describe how the new feature will be implemented.-->
+
+This SIP proposes to create a new version of the `EternalStorage` contract that instead of being limited to a single `associatedContract` via the `State` mixin, it can support any number of contracts that wish to use it in a safe and segregated manner.
+
+All getters and setters from `EternalStorage` will be used, yet with an extra initial parameter, `contractName`. This property will be the mapping for the storage entries. For any setter function, the `msg.sender` must match the address that the Synthetix `AddressResolver` has for it. As such, this storage contract must manage a reference to the `AddressResolver` (and this can be done one-time using the `ReadProxyAddressResolver`).
+
+## Questions
+
+> 1.  **Should the getters be limited to only be read the contract as well?** This would prevent other internal Synthetix contracts from reading directly from the storage contract on behalf of another contract and would prevent third party contracts from reading these values on-chain. The cost is that this limitation may have unforeseen consequences for integrations down the line, moreover it is slightly more gas efficint to look up the storage directly, whereas the benefit is that the contracts themselves are the abstraction for viewing storage, and it could be problematic to allow third party contracts to expect the storage to be formatted a certain way, and break those expectations in future releases.
+> 2.  **How to handle future refactoring of contracts?** If we were to store data for `Issuer` say, and then we split out burning from `Issuer` into `Burner`, how could we reuse storage from `Issuer` in `Burner`?
+>
+>     1. Use contract-auth calls to migrate the data over to the new mapping, one key at a time. This is manageable in the case of settings and properties, but much more onerous for address-based keys such as how [Issuer uses EternalStorage](https://github.com/Synthetixio/synthetix/blob/v2.21.15/contracts/Issuer.sol#L104).
+>
+>     2. A potentially better solution (though it could get ugly) is having an available contract mapping. Initially the mapping is empty but it could be added to by a contract that allows other named contracts in the `AddressResolver` to set/get on its behalf. So for the above example, `Issuer` would have something added to itself that says `Storage.addContractMapping("Issuer", "Burner")` that would allow `Burner` to pass through `Issuer` as a key and still have write access to that space. This isn't great because then `Burner` needs to keep a reference to the `"Issuer"` storage key, but is manageable. Any other suggestions?
```

Comment:
Maybe it's a bit grotesque, but it would be possible to manage our own hashes in order to simulate an indirection, something like this:

```
mapping(address => mapping(string => bytes32)) hashes;
mapping(bytes32 => ValueType) values;

function getValue(string variableName) external returns (ValueType) {
    return values[hashes[msg.sender][variableName]];
}

function getValue(string variableName, ValueType value) {
    values[hashes[msg.sender][variableName]] = value;
}

function addVariable(string variableName, address owner) {
    hashes[owner][variableName] = keccak256(abi.encodePacked(owner, variableName));
    // In an actual implementation, other functions would check that the hash actually exists before using it.
}

function grantVariableAccess(string variableName, address owner, address accessor) {
    // This will overwrite the hash in the accessor's own namespace, so contracts can't have a local and remote variable with the same name.
    hashes[accessor][variableName] = hashes[owner][variableName];
}
```

Reply:
Yeah I think that could work.  Agreed a bit grotesque but could be worth it. Though keying off `msg.sender` would mean invoking `grantAccess` each upgrade for each contact. Still, i can probably work something off of this.

Reply:
This is where I've landed (pseudo-code for values following your example):

```solidity
    mapping(bytes32 => mapping(bytes32 => ValueType)) values;
    mapping(bytes32 => bytes32) hashes;
    
    function getValue(bytes32 contractName, bytes32 record) external view returns (ValueType) {
        return values[hashes[contractName]][record];
    }

    function setValue(
        bytes32 contractName,
        bytes32 record,
        ValueType value
    ) external onlyContract(contractName) {
        if (hashes[contractName] == bytes32(0)) {
            // set to hash of sender and the labelled contract
            hashes[contractName] = keccak256(abi.encodePacked(msg.sender, contractName));
        } 
        values[hashes[contractName]][record] = value;
    }

    function deleteValue(bytes32 contractName, bytes32 record) external onlyContract(contractName) {
        delete UIntStorage[hashes[contractName]][record];
    }

    function migrateKey(bytes32 fromContractName, bytes32 toContractName) external onlyContract(fromContractName) {
        hashes[toContractName] = hashes[fromContractName];
    }

    modifier onlyContract(bytes32 contractName) {
        address callingContract = resolver.requireAndGetAddress(contractName, "Cannot find contract in Address Resolver");
        require(callingContract == msg.sender, "Can only be invoked by the configured contract");
        _;
    }
```

> Note: By not using `msg.sender` in the getters, we allow anyone to read (see Resolved Questions #1). As such, it's preferable to pass through the known contract name.

So for example if `Issuer` were to start setting storage for records `A_1` and `A_2` for instance, then the hash for the mapping would be `keccak256(address(Issuer), "Issuer")`. Then if we need to migrate the storage of `Issuer` to `Burner` say, we'd update the `Issuer` contract with:

```solidity
function migrate(bytes32 other) external onlyOwner {
        _storage.migrateKey(name, other);
}
```

then invoke `Issuer.migrate("Burner")` and invoke `AddressResolver.importAddresses(["Burner"], [address(Burner)})`. Successive reads from either `Issuer` or `Burner` will work, but then only `Burner` can call `set` or `delete`.

Any concerns with this approach?